In [17]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/logfile2-0/df.csv


In [18]:
!pip install --upgrade scikit-learn

In [19]:
df = pd.read_csv('/kaggle/input/logfile2-0/df.csv')

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 378837 entries, 0 to 378836
Data columns (total 18 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   type         378837 non-null  int64  
 1   subtype      378837 non-null  int64  
 2   level        378837 non-null  int64  
 3   srccountry   378837 non-null  int64  
 4   srcintf      378837 non-null  int64  
 5   srcintfrole  378837 non-null  int64  
 6   dstintf      378837 non-null  int64  
 7   dstintfrole  378837 non-null  int64  
 8   action       378837 non-null  int64  
 9   proto        378837 non-null  float64
 10  service      378837 non-null  int64  
 11  policyid     378837 non-null  float64
 12  appcat       378837 non-null  int64  
 13  duration     378837 non-null  float64
 14  sentbyte     378837 non-null  float64
 15  rcvdbyte     378837 non-null  float64
 16  crscore      378837 non-null  float64
 17  threat       378837 non-null  int64  
dtypes: float64(6), int64(12)

In [21]:
df

type  subtype  level  srccountry  srcintf  srcintfrole  dstintf  \
0          0        0      0           0        0            0        0   
1          0        0      0           0        0            0        0   
2          0        0      0           1        1            1        1   
3          0        0      0           0        0            0        0   
4          0        0      0           1        1            1        1   
...      ...      ...    ...         ...      ...          ...      ...   
378832     0        0      0          16        0            0        0   
378833     0        0      0           0        0            0        0   
378834     0        0      0           2        0            0        0   
378835     0        0      0           0        0            0        0   
378836     0        0      0           0        0            0        0   

        dstintfrole  action  proto  service  policyid  appcat  duration  \
0                 0       0    6.0        0      49.0       0      13.0   
1                 0       0    6.0        0      59.0       0      14.0   
2                 1       1    6.0        0      57.0       1       1.0   
3                 0       0    6.0        0      49.0       0      24.0   
4                 1       1    6.0        0      57.0       1       1.0   
...             ...     ...    ...      ...       ...     ...       ...   
378832            0       2    1.0        3      49.0       0      70.0   
378833            0       1    6.0        0      60.0       0       3.0   
378834            0       0    6.0        0      67.0       0       6.0   
378835            0       0    6.0        0      49.0       0      19.0   
378836            0       0    6.0        1      63.0       0       6.0   

        sentbyte  rcvdbyte  crscore  threat  
0         1097.0    9182.0      0.0       0  
1          424.0    8589.0      0.0       0  
2         1728.0    5255.0      0.0       0  
3          356.0    6708.0      0.0       0  
4         1728.0    5175.0      0.0       0  
...          ...       ...      ...     ...  
378832     132.0     172.0      0.0       0  
378833     416.0    9096.0      0.0       0  
378834     216.0     248.0      0.0       0  
378835     320.0    2530.0      0.0       0  
378836     124.0     244.0      0.0       0  

[378837 rows x 18 columns]

In [22]:
df.threat.value_counts()

threat
0    361756
2     10576
1      6500
3         5
Name: count, dtype: int64

# Binary Classification using Anomaly Detection

First, we will use Isolation Forest algorithm for anomaly detection on a scaled version of the dataset, in order to identify anomalies or malicious logs.

In [23]:
X=df.drop('threat',axis=1)
y=df['threat'].apply(lambda x: 0 if x==0 else 1)
X

type  subtype  level  srccountry  srcintf  srcintfrole  dstintf  \
0          0        0      0           0        0            0        0   
1          0        0      0           0        0            0        0   
2          0        0      0           1        1            1        1   
3          0        0      0           0        0            0        0   
4          0        0      0           1        1            1        1   
...      ...      ...    ...         ...      ...          ...      ...   
378832     0        0      0          16        0            0        0   
378833     0        0      0           0        0            0        0   
378834     0        0      0           2        0            0        0   
378835     0        0      0           0        0            0        0   
378836     0        0      0           0        0            0        0   

        dstintfrole  action  proto  service  policyid  appcat  duration  \
0                 0       0    6.0        0      49.0       0      13.0   
1                 0       0    6.0        0      59.0       0      14.0   
2                 1       1    6.0        0      57.0       1       1.0   
3                 0       0    6.0        0      49.0       0      24.0   
4                 1       1    6.0        0      57.0       1       1.0   
...             ...     ...    ...      ...       ...     ...       ...   
378832            0       2    1.0        3      49.0       0      70.0   
378833            0       1    6.0        0      60.0       0       3.0   
378834            0       0    6.0        0      67.0       0       6.0   
378835            0       0    6.0        0      49.0       0      19.0   
378836            0       0    6.0        1      63.0       0       6.0   

        sentbyte  rcvdbyte  crscore  
0         1097.0    9182.0      0.0  
1          424.0    8589.0      0.0  
2         1728.0    5255.0      0.0  
3          356.0    6708.0      0.0  
4         1728.0    5175.0      0.0  
...          ...       ...      ...  
378832     132.0     172.0      0.0  
378833     416.0    9096.0      0.0  
378834     216.0     248.0      0.0  
378835     320.0    2530.0      0.0  
378836     124.0     244.0      0.0  

[378837 rows x 17 columns]

We will use MinMaxScaler to scale the dataset.

In [24]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
import pickle

pickle.dump(MinMaxScaler().fit(X), open('scaler.pkl', 'wb'))
X = pd.DataFrame(MinMaxScaler().fit_transform(X))

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
X_train

0    1    2         3         4    5    6    7      8         9   \
223824  0.0  0.0  0.0  0.010870  0.333333  1.0  1.0  1.0  0.125  0.108696   
102622  0.0  0.0  0.0  0.184783  0.000000  0.0  0.0  0.0  0.250  0.000000   
329068  0.0  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.000  0.108696   
273470  0.0  0.0  0.0  0.010870  0.333333  1.0  1.0  1.0  0.125  0.108696   
155576  0.0  0.0  0.0  0.021739  0.000000  0.0  0.0  0.0  0.250  0.000000   
...     ...  ...  ...       ...       ...  ...  ...  ...    ...       ...   
51738   0.0  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.000  0.108696   
125546  0.0  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.000  0.108696   
322930  0.0  0.0  0.0  0.021739  0.000000  0.0  0.0  0.0  0.000  0.108696   
231577  0.0  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.125  0.108696   
109542  0.0  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.125  0.108696   

              10        11   12        13        14            15   16  
223824  0.000000  0.633333  1.0  0.000017  0.000949  1.068435e-04  0.0  
102622  0.003212  0.555556  0.0  0.000997  0.000022  3.834478e-07  0.0  
329068  0.000000  0.655556  0.0  0.001147  0.000265  5.531235e-05  0.0  
273470  0.000000  0.633333  1.0  0.000017  0.000925  5.673963e-05  0.0  
155576  0.003212  0.377778  0.0  0.000997  0.000022  3.834478e-07  0.0  
...          ...       ...  ...       ...       ...           ...  ...  
51738   0.000000  0.655556  0.0  0.000399  0.000178  5.888054e-05  0.0  
125546  0.000000  0.544444  0.0  0.000100  0.000168  2.258082e-06  0.0  
322930  0.000000  0.744444  0.0  0.000100  0.000131  2.641529e-06  0.0  
231577  0.000000  0.522222  0.0  0.000017  0.000435  1.405975e-06  0.0  
109542  0.000000  0.522222  1.0  0.002028  0.000195  9.063215e-05  0.0  

[303069 rows x 17 columns]

We will use Isolation Forest algorithm for anomaly detection. We will use n_estimators=800, bootstrap=True as hyperparameters.

In [25]:
from sklearn.ensemble import IsolationForest as iso

model = iso(n_estimators=800, bootstrap=True, random_state=69)
model.fit(X_train)

y_pred = model.predict(X_test)
y_pred

array([ 1, -1,  1, ...,  1,  1,  1])

In [26]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

y_pred = pd.Series(y_pred)
y_pred = y_pred.apply(lambda x: 1 if x == -1 else 0)
accuracy_score(y_test, y_pred)

0.9220383275261324

We get an accuracy of 92.2% which is quite good given that we have a huge and imbalanced dataset. However, in order to further analyze the results, we will use classification report and confusion matrix.

In [27]:
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.92      0.96     72352
           1       0.37      1.00      0.54      3416

    accuracy                           0.92     75768
   macro avg       0.68      0.96      0.75     75768
weighted avg       0.97      0.92      0.94     75768

[[66445  5907]
 [    0  3416]]


We can see that the model is not under-delivering, in the sense that it is not letting any genuine threats pass through. However we can see an error rate of 8.8% which is okay, given the size and quality of the dataset. Hence it is desirable that the model produce this error in overdetecting the threats rather than underdetecting them.

---

# Exporting the Model

In [28]:
import pickle

pickle.dump(model, open('raksha_v3.0.pkl', 'wb'))